sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala

```diff
@@ -1514,4 +1496,100 @@ class StatisticsSuite extends StatisticsCollectionTestBase with TestHiveSingleto
+        checkNumBroadcastHashJoins(res, 0,
+          "sort merge join should be taken as threshold is smaller than table size")
+      }
+
+      withSQLConf(SQLConf.AUTO_BROADCASTJOIN_THRESHOLD.key -> s"${origSizeInBytes + 1}") {
+        val res = sql(s"SELECT * FROM $table t1, $table t2 WHERE t1.key = t2.key")
+        checkNumBroadcastHashJoins(res, 1,
+          "broadcast join should be taken as the threshold is greater than table size")
+      }
+
+      withSQLConf(
+        SQLConf.DESERIALIZATION_FACTOR_CALC_ENABLED.key -> "true",
+        SQLConf.AUTO_BROADCASTJOIN_THRESHOLD.key -> s"${origSizeInBytes + 1}") {
+        sql(s"ANALYZE TABLE $table COMPUTE STATISTICS")
+        checkDeserializationFactor(table, exists = true)
+        val newSizeInBytes = spark.table(table).queryExecution.optimizedPlan.stats.sizeInBytes
+        assert(2 * origSizeInBytes <= newSizeInBytes)
+        logInfo(s"sizeInBytes after applying deserFactor: $newSizeInBytes")
+        val res = sql(s"SELECT * FROM $table t1, $table t2 WHERE t1.key = t2.key")
+        checkNumBroadcastHashJoins(res, 0,
+          "sort merge join should be taken despite the threshold is greater than the table" +
+            "size as the deserialization factor is applied")
+      }
+
+      withSQLConf(
+        SQLConf.DESERIALIZATION_FACTOR_CALC_ENABLED.key -> "false",
+        SQLConf.AUTO_BROADCASTJOIN_THRESHOLD.key -> s"${origSizeInBytes + 1}") {
+        sql(s"ANALYZE TABLE $table COMPUTE STATISTICS")
+        checkDeserializationFactor(table, exists = true)
+        val res = sql(s"SELECT * FROM $table t1, $table t2 WHERE t1.key = t2.key")
+        checkNumBroadcastHashJoins(res, 0,
+        "sort merge join should be taken despite deserialization factor calculation is " +
+          "disabled as the old factor is reused")
+      }
+
+      withSQLConf(SQLConf.AUTO_SIZE_UPDATE_ENABLED.key -> "true") {
+        val catalogTableBefore = getCatalogTable(table)
+        val deserFactorBefore = catalogTableBefore.stats.get.deserFactor.get
+        sql(s"INSERT INTO TABLE $table PARTITION (ds='2010-01-02') SELECT '1', 'A' from SRC")
+        spark.sessionState.catalog.refreshTable(TableIdentifier(table))
+        val catalogTable1 = getCatalogTable(table)
+        assert(catalogTable1.stats.isDefined &&
+          catalogTable1.stats.get.deserFactor.isDefined)
+        assert(catalogTable1.stats.get.deserFactor.get === deserFactorBefore,
+          "deserFactor should not change by adding a smaller or same size partition")
+
+        sql(s"INSERT INTO TABLE $table PARTITION (ds='2010-01-03') SELECT * FROM SRC")
+        spark.sessionState.catalog.refreshTable(TableIdentifier(table))
+        val catalogTable2 = getCatalogTable(table)
+        assert(catalogTable2.stats.isDefined &&
+          catalogTable2.stats.get.deserFactor.isDefined)
+        assert(catalogTable2.stats.get.deserFactor.get > deserFactorBefore,
```

Comment:
The test implies that size of data impacts the deserFactor (ration), when in reality (I believe) it's the make-up of the data: Whether data is incrementing sequentially, or is repeating, or is random. Is that not the case?